src/curses.rs

```diff
@@ -441,33 +366,42 @@ impl Term {
 /// error output to stderr in case of failure.
 ///
 /// Any existing references from `curses::term()` will be invalidated by this call!
-pub fn setup<F>(term: Option<&CStr>, fd: i32, configure: F) -> Option<Arc<Term>>
+pub fn setup<F>(term: Option<&str>, configure: F) -> Option<Arc<Term>>
 where
     F: Fn(&mut Term),
 {
     // For now, use the same TERM lock when using `cur_term` to prevent any race conditions in
     // curses itself. We might split this to another lock in the future.
     let mut global_term = TERM.lock().expect("Mutex poisoned!");
 
-    let result = unsafe {
-        // If cur_term is already initialized for a different $TERM value, calling setupterm() again
-        // will leak memory. Call del_curterm() first to free previously allocated resources.
-        let _ = sys::del_curterm(get_curterm());
-
-        let mut err = 0;
-        if let Some(term) = term {
-            sys::setupterm(term.as_ptr(), fd, &mut err)
+    let res = if let Some(term) = term {
+        terminfo::Database::from_name(term)
+    } else {
+        // For historical reasons getting "None" means to get it from the environment.
+        terminfo::Database::from_env()
+    }
+    .or_else(|x| {
+        // Try some more paths
+        let t = if let Some(term) = term {
+            term.to_string()
+        } else if let Ok(name) = env::var("TERM") {
+            name
         } else {
-            sys::setupterm(core::ptr::null(), fd, &mut err)
-        }
-    };
+            return Err(x);
+        };
+        let mut path = PathBuf::from("/usr/pkg/share/terminfo");
```

Comment:
if this is a NetBSD-specific workaround then maybe enable it only there?
Also, maybe this should use `env!("PREFIX")` instead of hardcoding `/usr`?
The only place where we hardcode `/usr` is in the `DEFAULT_PATH` but PREFIX comes first there.

Reply:
> Also, maybe this should use env!("PREFIX") instead of hardcoding /usr?

This path is not dependent on where fish is installed, so using PREFIX would be wrong. I am unsure if pkgsrc can be set to use a different path, but this is the default one.

> if this is a NetBSD-specific workaround then maybe enable it only there?

This is the path used on NetBSD if you install from pkgsrc (the default package manager there), but pkgsrc itself can also be used on other platforms.

More than that, this is just one more path to search, if we find a valid terminfo entry there we should use it regardless of platform.

Reply:
ok that sounds fine. Would be nice to get rid of this workaround eventually

Reply:
We can, of course, also propose this [upstream](https://github.com/meh/rust-terminfo).

Unlike the FreeBSD path, I didn't because this one is a bit less generic and I wanted to see what looking up custom paths would look like, in case it turned out that we would have to add e.g. build options to add more search paths.

Reply:
Or propose a method to terminfo that allows passing a list of directories to search.

Reply:
`man terminfo` says that after checking `TERMINFO_DIRS` etc it does:

> •   Finally, ncurses searches these compiled-in locations:
>    •   a list of directories (no default value), and
>    •   the system terminfo directory, /usr/share/terminfo (the compiled-in default).

I guess "(no default value)" means it's platform specific and we have to reimplement that?

Reply:
I would assume that is referring to the [--with-terminfo-dirs](https://invisible-island.net/ncurses/INSTALL.html#option:with-terminfo-dirs) ncurses build option.

Technically, distributions *could* set that one, I doubt many do.

We already have checked a *bunch* of systems here, and we already have them covered.

Reply:
Note: Debian [uses --with-terminfo-dirs](https://salsa.debian.org/debian/ncurses/-/blob/b42033ffb63582d99294f1aea351e3dbcfa1e94c/debian/rules#L136), but to add "/etc/terminfo:/lib/terminfo:/usr/share/terminfo", all of which [are already checked by rust-terminfo](https://github.com/meh/rust-terminfo/blob/870327dd79beaecf50db09a15931d5ee1de2a24d/src/database.rs#L180-L182). Likewise [for fedora](https://src.fedoraproject.org/rpms/ncurses/blob/rawhide/f/ncurses.spec#_146). [NixOS](https://github.com/NixOS/nixpkgs/blob/bc79dc30fdf28f42fadea0a62529ee589b8e0824/pkgs/development/libraries/ncurses/default.nix#L50) adds "/run/current-system/sw/share/terminfo" (tho that seems to only compile it in for static binaries?).

----

I am inclined to merge this as-is, with the search path it has, and if anyone needs other paths we either add them or, worst case, add a build option to add more paths.